consensus/types/src/light_client_header.rs

```diff
@@ -53,6 +54,85 @@ pub struct LightClientHeader<E: EthSpec> {
     pub phantom_data: PhantomData<E>,
 }
 
+impl<E: EthSpec> LightClientHeader<E> {
+    pub fn block_to_light_client_header(
+        block: &SignedBeaconBlock<E>,
+        fork_name: ForkName,
+    ) -> Result<Self, Error> {
+        let header = match fork_name {
+            ForkName::Base => return Err(Error::AltairForkNotActive),
+            ForkName::Altair | ForkName::Merge => LightClientHeader::Altair(
+                LightClientHeaderAltair::block_to_light_client_header(block)?,
+            ),
+            ForkName::Capella => LightClientHeader::Capella(
+                LightClientHeaderCapella::block_to_light_client_header(block)?,
+            ),
+            ForkName::Deneb => LightClientHeader::Deneb(
+                LightClientHeaderDeneb::block_to_light_client_header(block)?,
+            ),
+        };
+        Ok(header)
+    }
+
+    pub fn from_ssz_bytes(bytes: &[u8], fork_name: ForkName) -> Result<Self, ssz::DecodeError> {
+        let header = match fork_name {
+            ForkName::Altair | ForkName::Merge => {
+                let header = LightClientHeaderAltair::from_ssz_bytes(bytes)?;
+                LightClientHeader::Altair(header)
+            }
+            ForkName::Capella => {
+                let header = LightClientHeaderCapella::from_ssz_bytes(bytes)?;
+                LightClientHeader::Capella(header)
+            }
+            ForkName::Deneb => {
+                let header = LightClientHeaderDeneb::from_ssz_bytes(bytes)?;
+                LightClientHeader::Deneb(header)
+            }
+            ForkName::Base => {
+                return Err(ssz::DecodeError::BytesInvalid(format!(
+                    "LightClientHeader decoding for {fork_name} not implemented"
+                )))
+            }
+        };
+
+        Ok(header)
+    }
+}
+
+impl<E: EthSpec> Encode for LightClientHeader<E> {
```

Comment:
I think we can just derive `Encode` without needing a manual impl?

Reply:
When deriving `Encode` I get a `not all enum variants are variably-sized` error message when running `ssz_tests_by_fork!`. The only work around I've found so far is manually implementing `Encode`

Reply:
ah right, the limitation here:
https://github.com/sigp/ethereum_ssz/blob/649ed02ef19c6b2c364d0f4b514cc1c14f6b44d0/ssz_derive/src/lib.rs#L530-L534

The Altair variant is fixed sized, so I think you're right that we might have to manually implement this.